shared-libs/lineage/src/lineage.js

```diff
@@ -0,0 +1,403 @@
+        .map(function(row) {
+          return row.doc;
+        })
+        .filter(function(doc) {
+          return !!doc;
+        });
+    });
+};
+
+var hydrateParents = function(docs, parents) {
+  if (!parents || !parents.length) {
+    return docs;
+  }
+
+  var findById = function(id, docs) {
+    return docs.find(function(doc) {
+      return doc._id === id;
+    });
+  };
+
+  docs.forEach(function(doc) {
+    var current = doc;
+    if (doc.type === 'data_record') {
+      var contactDoc = findById(current.contact._id, parents);
+      if (contactDoc) {
+        doc.contact = contactDoc;
+      }
+      current = doc.contact;
+    }
+
+    while (current) {
+      if (current.parent && current.parent._id) {
+        var parentDoc = findById(current.parent._id, parents);
+        if (parentDoc) {
+          current.parent = parentDoc;
+        }
+      }
+      current = current.parent;
+    }
+  });
+  return docs;
+};
+
+var hydrateLeafContacts = function(docs, contacts) {
+  var subDocsToHydrate = [];
+  docs.forEach(function(doc) {
+    var current = doc;
+    if (doc.type === 'data_record') {
+      current = doc.contact;
+    }
+    while (current) {
+      subDocsToHydrate.push(current);
+      current = current.parent;
+    }
+  });
+  fillContactsInDocs(subDocsToHydrate, contacts);
+  return docs;
+};
+
+var hydratePatient = function(doc) {
+  return fetchPatientLineage(doc).then(function(patientLineage) {
+    if (patientLineage.length) {
+      var patientDoc = patientLineage.shift();
+      fillParentsInDocs(patientDoc, patientLineage);
+      doc.patient = patientDoc;
+    }
+    return doc;
+  });
+};
+
+var hydratePatients = function(docs) {
+  return Promise.all(docs.map(hydratePatient)).then(function() {
+    return docs;
+  });
+};
+
+var hydrateDocs = function(docs) {
+  if (!docs.length) {
+    return Promise.resolve([]);
+  }
+  var parentIds = collectParentIds(docs);
+  var hydratedDocs = JSON.parse(JSON.stringify(docs));
+  return fetchDocs(parentIds)
+    .then(function(parents) {
+      hydrateParents(hydratedDocs, parents);
+      return fetchDocs(collectLeafContactIds(hydratedDocs));
+    })
+    .then(function(contacts) {
+      hydrateLeafContacts(hydratedDocs, contacts);
+      return hydratePatients(hydratedDocs);
+    });
+};
+
+// Minifies things you would attach to another doc:
+//   doc.parent = minify(doc.parent)
+// Not:
+//   minify(doc)
+var minifyLineage = function(parent) {
+  if (!parent || !parent._id) {
+    return parent;
+  }
+
+  var result = { _id: parent._id };
+  var minified = result;
+  while (parent.parent && parent.parent._id) {
+    minified.parent = { _id: parent.parent._id };
+    minified = minified.parent;
+    parent = parent.parent;
+  }
+  return result;
+};
+
+module.exports = {
+  init: function(dependencies) {
+    dependencies = dependencies || {};
+    Promise = dependencies.Promise;
+    DB = dependencies.DB;
+    return module.exports;
+  },
```

Comment:
I'm not a fan of this pattern because it's hoisting `Promise` and `DB` up in scope so they will hang around too long and pollute other scopes.

Instead can we...

either 1) Have init return an object with `DB` and `Promise` properties and all the functions below? eg:

```js
module.exports = function(db, promise) {
  var fetchHydratedDoc = function() { ... };
  return { fetchHydratedDoc: fetchHydratedDoc };
};
```

or 2) Pass `promise` and `db` in to each function as required, eg:

```js
module.exports = {
  fetchHydratedDoc: function(db, id) { ... };
};
```

Option 1 is a bit cleaner in implementation but I think option 2 is cleaner in API because it's clear what each function uses. Right now it's confusing which dependencies you need to pass in to the init function to do what you want it to do. Option 2 also makes mocking trivial. I'm leaning towards option 2 but open to ideas...

Reply:
Open 1 doesn't allow mocking, which is why it was done this way. Option 2 goes against the entire rest of our codebase, and would mean really awkward stuff like passing Promise around.

The current code is a not great compromise, but I'm not sure of how to improve it.

> I'm not a fan of this pattern because it's hoisting Promise and DB up in scope so they will hang around too long and pollute other scopes.

Could you explain this one? I don't understand what you mean, which other scopes?